core/src/main/java/be/orbinson/aem/dictionarytranslator/services/impl/CombiningMessageEntryResourceProvider.java

```diff
@@ -67,45 +64,63 @@ private Map<String, Object> getValuesAndMessageEntryPaths(Resource dictionaryRes
             return null;
         }
 
-        String key = Text.getName(path);
-        String dictionaryPath = Text.getRelativeParent(path, 1).replaceFirst(ROOT, "");
+        String key = extractKeyFromPath(path);
+        String dictionaryPath = getDictionaryPath(path);
         Resource dictionaryResource = resourceResolver.getResource(dictionaryPath);
 
-        if (dictionaryResource != null && isMessageEntry(dictionaryResource, key)) {
-            Map<String, Object> properties = new HashMap<>();
-            properties.put(KEY, key);
-            properties.put("path", path);
-            properties.put("editable", dictionaryService.isEditableDictionary(dictionaryResource));
-            properties.put(DICTIONARY_PATH, dictionaryPath);
+        if (dictionaryResource != null) {
+            @NotNull Map<String, Message> messagePerLanguage = new HashMap<>();
             List<String> languages = dictionaryService.getLanguages(dictionaryResource);
-            properties.put(LANGUAGES, languages);
-            properties.putAll(getValuesAndMessageEntryPaths(dictionaryResource, key, languages));
-            return new ValueMapResource(resourceResolver, path, RESOURCE_TYPE, new ValueMapDecorator(properties));
+            for (String language : languages) {
+                Message message;
+                try {
+                    message = dictionaryService.getMessages(dictionaryResource, language).get(key);
+                } catch (DictionaryException e) {
+                    throw new ResourceNotFoundException("Unable to get message entries for language '" + language + "' in dictionary '" + dictionaryPath + "'", e);
+                }
+                messagePerLanguage.put(language, message);
+            }
+           return createResource(parent, resourceResolver, path, dictionaryService.isEditableDictionary(dictionaryResource), messagePerLanguage);
+        } else {
+            throw new ResourceNotFoundException(path, "Unable to get underlying dictionary resource for path '" + dictionaryPath + "'");
         }
-        return null;
     }
 
-    private boolean isMessageEntry(Resource dictionaryResource, String key) {
-        List<String> languages = dictionaryService.getLanguages(dictionaryResource);
-        // in order to speed things up always start with language "en" (if existing)
-        String mostCompleteLanguage = Locale.ENGLISH.getLanguage();
-        if (languages.contains(mostCompleteLanguage)) {
-            languages.remove(mostCompleteLanguage);
-            languages.add(0, mostCompleteLanguage);
-        }
+    static @NotNull Resource createResource(@Nullable Resource parent, @NotNull ResourceResolver resourceResolver, 
```

Comment:
this is a bit ambiguous for me as method name, with the actual `create` method that exists: https://sling.apache.org/apidocs/sling9/org/apache/sling/spi/resource/provider/ResourceProvider.html#create-org.apache.sling.spi.resource.provider.ResolveContext-java.lang.String-java.util.Map-